Validate Input for Whitelist and Blacklist (Exact)

slavabez opened the pull request:
Addresses the #84 issue

Added client-side validation using the regex from the API. When `<DomainInput>` has invalid input, adds the is-invalid class to make it have a red outilne, and green if valid.

Uses the `<Alert>` from the parent component, communicates through a prop function.

Didn't see where to write tests for these, let me know where and I'll add some

slavabez commented:
My bad, will fix the tests

slavabez commented:
Thanks for the review! I'll add all the changes. I think the project should consider using prettier so that all the parentheses, indents and all that kind of stuff was automatic

Mcat12 commented:
I agree, some kind of formatter and linter should be added.
The API already uses rustfmt, and I plan on eventually adding clippy support (linting). It has helped minimize style review comments in that repo.

slavabez commented:
I'd be happy to add prettier as a separate issue/pr, it's very simple to install and has nice defaults. I'll do the other changes later today

slavabez commented:
Addressed all the changes! Was more work than I anticipated :D 
Most things as requested, except re-named some variables and prop names to better reflect what they do now.
Changed the `DomainInput` to a form, eliminating the need to an Enter key listener, and the Add button type to submit. Required changing the tests a little, to simulate form submission rather than button clicking (I believe button clicking to submit a form isn't supported with Enzyme's shallow rendering)

slavabez commented on src/components/list/DomainInput.js:
```diff
@@ -11,21 +11,34 @@
   state = {
-    domain: ""
+    domain: "",
+    isValid: true
```
Changed as requested, no more on the fly checking, only when submitted

Mcat12 commented on src/validate.js:
```diff
@@ -0,0 +1,46 @@
+*  Network-wide ad blocking via your own hardware.
+*
+*  Web Interface
+*  Various utilities including the API
```
Update the description.

Mcat12 commented on src/validate.js:
```diff
@@ -0,0 +1,46 @@
+    const segments = hostname.split(".");
+    // If at least one segment is empty or longer than 63 chars - sets to true, if none - false
+    const hasLongSegments = segments.some(
+      segment => !segment || segment.length > 63
```
Why is `!segment` necessary?

slavabez replied:
The idea was to check for empty segments as well as segments that are too long

Mcat12 commented on src/validate.js:
```diff
@@ -0,0 +1,46 @@
+*  This file is copyright under the latest version of the EUPL.
+*  Please see LICENSE file for your rights under this license. */
+
+export const validate = {
```
Flatten the exports from this file (export each function individually).
You should also rename the functions then to something like `isValidDomain`.

Mcat12 commented on src/components/list/ListPage.js:
```diff
@@ -98,10 +98,17 @@ class ListPage extends Component {
+      this.setState({
+        errorMsg: err.message
+      });
+    } else {
```
Also note: the message should be wrapped in a call to `this.props.t` so it gets translated. This is one reason why it might be better to put the error message in `ListPage` instead.

slavabez commented:
Thanks for the review! I'll get the changes committed quick

slavabez commented:
Done, made the changes. Sorry for making so much work for you, hope it's worth it :smile:

Mcat12 commented on src/validate.js:
```diff
@@ -0,0 +1,43 @@
+  // If the hostname without periods make a number, deny
+  if (isStrictNumeric(joined)) return false;
+
+  return /([a-zA-Z0-9]+(-[a-zA-Z0-9]+)*)+(\.([a-zA-Z0-9]+(-[a-zA-Z0-9]+)*))*$/.test(
```
Add `^` to the start of the regex so it forces the whole string to match (this exists in the API code).

A good test for this would be `&.domain.com`

slavabez replied:
Good catch! My inexperience with regex is showing :smile:

Mcat12 commented on src/components/list/DomainInput.js:
```diff
@@ -12,39 +12,49 @@ import React, { Component } from 'react';
+    e.preventDefault();
 
+    const domain = this.state.domain;
+    const isValid = isValidDomain(domain);
```
`ListPage` should take in a validation function which it gives to `DomainInput`, so that the regex page can have its own special validator. Then, the error message should also be extracted out to be a prop on `ListPage`.

Mcat12 commented:
Thanks for quickly adding the requested changes!

slavabez commented:
Cool! I'll get these done at some point during the weekend

Mcat12 commented on public/i18n/en/lists.json:
```diff
@@ -8,5 +8,6 @@
-}
\ No newline at end of file
+    "Input a regular expression": "Input a regular expression",
+    "Invalid domain error": "Not valid domain format (use example.com or sub.example.com)"
```
There is a script for updating the strings and translations, `npm run update-translations`. It requires the `POEDITOR_API_TOKEN` environment variable to be set to your POEditor API token. I can add the string and update the branch with the translations, if you don't have a POEditor account.

Also, the string makes a bit more sense as "Not a valid domain", because example.com and sub.example.com are already in the placeholder. I'll add this string to POEditor.

slavabez replied:
Yeah, please add the strings once the PR is absorbed

slavabez commented on src/validate.test.js:
```diff
@@ -0,0 +1,121 @@
+    });
+
+    describe("should fail", () => {
+      it("rejects '&.domain.com'", () => {
```
Added another test case to address the above regex change

Mcat12 commented on src/components/list/DomainInput.js:
```diff
@@ -15,53 +15,64 @@ import { api } from "../../utils";
   onEnter: PropTypes.func.isRequired,
-  onRefresh: PropTypes.func.isRequired
+  onRefresh: PropTypes.func.isRequired,
+  validationFunction: PropTypes.func.isRequired,
```
Rename to `isValid`

slavabez commented:
Done! Renamed the `validationFunction` to `isValid`

Mcat12 commented on src/components/list/ListPage.js:
```diff
@@ -121,7 +128,9 @@ class ListPage extends Component {
           onEnter={this.onEnter}
-          onRefresh={this.onRefresh}/>
+          onRefresh={this.onRefresh}
+          isValid={isValidDomain}
```
This still only lets the regex list take in domains. Move the `isValid` method to be a prop on `ListPage`, and for the `Regexlist` pass in `() => true`. This will make implementing #85 easier.

slavabez commented:
Like this? Moved the isValdi to be compulsory prop of ListPage, and modified tests to remove warnings

Mcat12 commented:
Now there's just a conflict with `package-lock.json` that needs to be resolved. You should take `development`'s version, which came from #89.

slavabez commented:
Done, hopefully the tests pass :smile:

slavabez commented:
I'll try that!

slavabez commented:
I think that's done it

Mcat12 commented:
Sorry, I already did it (see the previous commit)
